File: euler/c++/P549.cc
/*
    Divisibility of Factorials
    Problem 549
*/

#include <iostream>

const uint64_t LOWER_LIMIT = 2;
const uint64_t UPPER_LIMIT = 100000000;

int main(int argc, char * argv[]) {

    // Initialize the minimum value of each to zero. We then iterate over each
    // value and set it to the correct one.
    uint64_t * smallest = new uint64_t[UPPER_LIMIT + 1]();
    for(uint64_t prime = LOWER_LIMIT; prime <= UPPER_LIMIT; prime++) {
        // If the value is not zero, it's not a prime and so we skip it
        if(smallest[prime]) continue;
        uint64_t c_power = 1, c_value = prime;
        uint64_t p_found = 1, minimum = prime;

        while(c_value <= UPPER_LIMIT) {
            // Get the new minimum value given the power increase
            while(c_power > p_found) {
                uint64_t new_factor = (minimum += prime);
                do p_found++;
                while((new_factor /= prime) % prime == 0);
            }

            // Flag all multiples of the prime power with the new value
            for(uint64_t j = c_value; j <= UPPER_LIMIT; j += c_value)
                smallest[j] = smallest[j] < minimum ? minimum : smallest[j];
            c_power++;
            c_value *= prime;
        }
    }

    // Calculate and print the sum
    uint64_t total_sum = 0;
    for(uint64_t i = 0; i <= UPPER_LIMIT; i++) total_sum += smallest[i];
    delete [] smallest;
    std::cout << total_sum << std::endl;

    return 0;
}

File: euler/python/P549.py
"""
Divisibility of Factorials
Problem 549
"""

LOWER_LIMIT = 2
UPPER_LIMIT = 100

smallest = [0] * (UPPER_LIMIT + 1)
for prime in xrange(LOWER_LIMIT, UPPER_LIMIT + 1):
    # If the value is still zero, that means that the number is prime, so we
    # iterate with it and all of its powers.
    if smallest[prime] == 0:

        # Get each power of the prime and iterate through the list, updating
        # the value to whatever number is required to get that power.
        current_power, current_value = 1, prime

        # Keep tracks of the copies of the prime found so far in the factorial.
        primes_found, minimum = 1, prime

        # Go through the whole sieve
        while current_value <= UPPER_LIMIT:

            # Before sieving, we make sure that the current power we have can
            # divide our minimum factorial.
            while current_power > primes_found:

                # If it can't, we keep adding prime to the minimum factorial
                # until we have enough factors.
                minimum += prime

                # Check how many new prime factors the number has. Note that
                # we don't need to check the ones between minimum
                # and minimum + prime since there cannot be any.
                new_factor = minimum
                while new_factor % prime == 0:
                    primes_found += 1
                    new_factor /= prime

            # Go through each multiple of the power and make sure the number
            # is at least equal to our minimum
            for j in xrange(current_value, UPPER_LIMIT + 1, current_value):
                smallest[j] = max(smallest[j], minimum)

            # Go to the next prime power
            current_power += 1
            current_value *= prime

# Finally, sum all of the values
print sum(smallest)
